src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala

```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
+//  import actorSystem.dispatcher
+  import GcsAuthHelpers._
 
-object GcsAuthorizationProvider extends AuthorizationProvider with LazyLogging {
-  implicit val actorSystem = ActorSystem("agora")
-  import actorSystem.dispatcher
+  override def namespaceAuthorization(entity: AgoraEntity, username: String): AgoraPermissions = {
+    val bucketNotFound = checkBucketExistence(entity, username)
 
-  val timeout = 20 seconds
+    if (bucketNotFound) return new AgoraPermissions(Create)
+    else getBucketPermissions(entity, username)
+  }
 
-  val gcsBaseUrl = "https://www.googleapis.com/storage/v1/b/"
+      private def checkBucketExistence(entity: AgoraEntity, username: String): Boolean = {
+        val url = bucketUrl(entity)
+        val bucket = gcsGET(url)
 
-  // GCS sends back a malformed ETag header according to HTTP standards. Spray complains. We currently don't care about ETag, so filter it out.
-  val removeEtagHeaders: HttpResponse => HttpResponse =
-    r => r.withHeaders(r.headers.filter(!_.name.startsWith("ETag")))
+        val bucketNotFound = bucket map { bucketResponse =>
+          bucketResponse.status == NotFound
+        }
+        Await.result(bucketNotFound, timeout)
+      }
 
-  val pipeline = addCredentials(OAuth2BearerToken(GoogleCredentialHandler.accessToken)) ~> sendReceive ~> removeEtagHeaders
 
-  private case class GoogleResponse(entity: AgoraEntity, response: HttpResponse)
+      private def getBucketPermissions(entity: AgoraEntity, username: String): AgoraPermissions = {
+        val url = bucketAclUrl(entity, username)
+        val bucketAcl = gcsGET(url)
+
+        val permissions = bucketAcl map { acl =>
+          val role = GcsBucketRole(unrollGoogleAcl(acl))
+          RoleTranslator.gcsBucketToNamespacePermissions(role)
+        }
+        Await.result(permissions, timeout)
+      }
+
+  override def entityAuthorization(entity: AgoraEntity, username: String): AgoraPermissions = {
+    val url = objectAclUrl(entity, username)
+    val objectAcl = gcsGET(url)
 
-  def namespaceToBucketName(namespace: String): String = "agoranamespace_" + namespace.toLowerCase
+    val permissions = objectAcl.map { acl =>
+      val role = GcsObjectRole(unrollGoogleAcl(acl))
+      RoleTranslator.gcsObjectToEntityPermissions(role)
+    }
+    Await.result(permissions, timeout)
+  }
+
+  override def createEntityAuthorizations(entity: AgoraEntity, username: String): Unit = {
+    val bucket = Promise[HttpResponse]()
+    val bucketAcl = Promise[HttpResponse]()
+    val _object = Promise[HttpResponse]()
+    val objectAcl = Promise[HttpResponse]()
 
-  def entityToObjectName(entity: AgoraEntity): String = "agoraEntity_" + entity.entityType.get + "_" + entity.name.get + "_" + entity.snapshotId.get
+    createBucket(entity, username, bucket)
+
+    bucket.future onSuccess { case r =>
+      if (r.status != OK && r.status != Conflict) throw new Exception("Could not create bucket.")
```

Comment:
AddHandler does not seem to be looking for these exceptions, in order to complete request. Do they crash the system?